### PR TITLE
[WIP] Stats "First View"

### PR DESCRIPTION
**Pull request by mattsherman**
**WIP: Not ready for review or merging yet.**

This PR implements a "first view" screen for the `/stats` section.

Support for other sections will be possible once #6651 is completed (Reduxifying `<Main>`). 

Test live: https://calypso.live/?branch=add/stats-first-view-ui

### CODE REVIEW

File: client/components/first-view/index.jsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		const firstViewContainerClasses = classNames( 'first-view', {
+			active: this.state.showFirstView
+		} );
+
+		const mainViewContainerClasses = classNames( 'first-view-main-view', {
+			inactive: this.state.showFirstView
+		} );
+
+		return (
+			<div className="main main-column first-view-wrapper" role="main">
```
Comment (mtias) [inline client/components/first-view/index.jsx]: I know this is early and in progress, but before it gets too far in development, I think the approach needs a bit more attention. All these main classes should only be coming from using the `<Main>` component. I also think this could look more like:

```
<Main>
    ... my section stuff
    <FirstView> (using RootChild)
</Main>
```

In terms of structure that seems more straightforward and semantic — first view is a child of the main component. Do you think something like that could work?

Comment (mattsherman) [inline client/components/first-view/index.jsx]: Thanks for the feedback @mtias -- I will look into structuring things as you suggested and making use of RootChild.

Comment (mtias) [inline client/components/first-view/index.jsx]: Sure thing, thank you :)

**Comment by mattsherman** [comment]
@mtias This is still very much WIP, but I have taken your suggestion and changed things to go with a RootChild-based approach. What do you think of this?

Note: We are starting with a "first view" for stats, but eventually we will implement "first views" for other screens as well. So, having a `FirstViewable` container seems to make sense as well, to avoid implementation duplication in sections that want to support "first view".

File: client/my-sites/stats/stats-insights/index.jsx [hunk 2]
```diff
@@ -72,51 +81,72 @@ export default React.createClass( {
...
-									emailFollowersList={ emailFollowersList }
-									followList={ followList } />
-								<StatsModule
-									path={ 'publicize' }
-									moduleStrings={ moduleStrings.publicize }
-									site={ site }
-									dataList={ publicizeList } />
-							</div>
+		const content = (
+			<div id="my-stats-content">
```
Comment (mtias) [inline client/my-sites/stats/stats-insights/index.jsx]: This should be the `<Main>` component without id if possible. I don't think we need to move it all to a content variable either, I think it reads much better in the main render tree.

File: client/my-sites/stats/stats-insights/index.jsx [hunk 3]
```diff
@@ -72,51 +81,72 @@ export default React.createClass( {
...
+								dataList={ publicizeList } />
 						</div>
 					</div>
 				</div>
 			</div>
 		);
+
+		if ( config.isEnabled( 'stats/first-view' ) ) {
+			return (
+				<FirstViewable isFirstViewActive={ this.state.isFirstViewActive } onFirstViewDismiss={ this.dismissFirstView }>
```
Comment (mtias) [inline client/my-sites/stats/stats-insights/index.jsx]: I don't think first viewable should be a wrapper, I think it should be a child, because conceptually it's just one state of the component.

**Comment by mtias** [comment]
@mattsherman the root-child suggestion was primarily to avoid needing a wrapper component at all. I think this component should be a child of Main because conceptually it's just a different state it can be in.

```
<Main className="stats">
    <StatsNavigation section="insights" site={ site } />
    etc
    </FirstView>
</Main>
```

Otherwise you could consider a higher order component?

Wrapping all the views in `FirstViewable` seems quite noisy.

File: client/my-sites/stats/stats-insights/index.jsx (hunk 2 above)
Comment (mattsherman) [inline client/my-sites/stats/stats-insights/index.jsx]: I see. I wasn't aware that the stats section wasn't using the preferred components already. So, just to make sure I'm clear, this suggestion is irregardless of this particular PR (in other words, it would be recommended even if they scrapped the whole "first view" stuff).

Comment (mattsherman) [inline client/my-sites/stats/stats-insights/index.jsx]: If I go with the higher order component approach, this separate content variable will be unnecessary. I agree it reads better in the main render tree.

**Comment by mattsherman** [comment]
@mtias The main problem I'm trying to tackle is how to provide a component that handles the animation for both the `FirstView` and the `Main` component that the `FirstView` is a 'child' of. When the `FirstView` is shown, I also need to set CSS classes on the `Main` component in order to blur and scale it. And I want to provide this capability to a number of different `Main` components in the future (we're starting with stats, but other sections will get first views eventually). That's why I felt a wrapper component was necessary.

I wasn't aware of the "higher order component" approach. Is this what you are talking about? http://www.natpryce.com/articles/000814.html

Do you have any examples in Calypso that I should use as reference?

That seems like a clearer approach that still wraps things but avoids the noise you were concerned about. I will change things to take this approach.

Thanks for helping to make sure this PR does things the most appropriate way for Calypso -- I really appreciate the early feedback.

File: client/my-sites/stats/stats-first-view/index.jsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ * External dependencies
+ */
+import React from 'react';
+
+export default React.createClass( {
+	render() {
+		return (
+			<div>
+				{ /* TODO: serve this image from proper place */ }
+				<img src="/calypso/images/stats/illustration-stats.svg" />
```
Comment (mtias) [inline client/my-sites/stats/stats-first-view/index.jsx]: With React 15 we can inline the svgs and avoid a request, just render the `<svg>` directly.

**Comment by mattsherman** [comment]
Rebased against master to pull in latest and clean up commits, since implementation approach changed a few times.

Waiting on https://github.com/Automattic/wp-calypso/pull/6252 -- once that is in, proper Redux support for the state will be implemented and this will be ready for review.

In the mean time, the UI interaction can be tested.

**Comment by mattsherman** [comment]
Note: I will rename the concept of `dismiss` to `hide` in the UI code to match the Redux state implementation in #6252.

**Comment by jonathansadowski** [comment]
image
I noticed that the PN prompt shows up weird in this context

**Comment by mattsherman** [comment]
Good catch!

**Comment by shaunandrews** [comment]
- It took me a few to realize that the overlay is only shown on the Insights tab — it should be shown the first time a user views any Stats page. The overlay is related to the specific tab, but the sidebar item.
- Switching to Stats (or loading it for the first time), the overlay animates into view. It shouldn't. It should just "be there" already. The only transitions are for the exit.
- I think the "Don't show this again" option should be checked by default.
- The label and checkbox are mis-aligned:

screen shot 2016-06-29 at 3 41 12 pm

**Comment by mattsherman** [comment]
Thanks @shaunandrews – I'll take care of those!

(I even had to put in extra code to get it to animate when shown... Guess I can pull that out! 😄)

**Comment by shaunandrews** [comment]
-  After dismissing the overlay, I'm unable to click on anything else in the main area of the page. (The masterbar and sidebar still work)

**Comment by mattsherman** [comment]
Oops... I'll fix that!

File: client/components/enable-first-view/index.jsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+			const classes = classNames( 'first-viewable', {
+				'is-first-view-visible': this.state.isFirstViewVisible
+			} );
+
+			return (
+				<div className={ classes }>
+					<FirstView isVisible={ this.state.isFirstViewVisible }
+							onHide={ this.hideFirstView }
+							isEnabledNextTime={ this.props.isFirstViewEnabledNextTime }
+							onEnableOrDisableNextTime={ this.enableOrDisableFirstViewNexTime }>
```
Comment (mjuhasz) [inline client/components/enable-first-view/index.jsx]: There's a typo here: a missing 't' in `enableOrDisableFirstViewNexTime `.

**Comment by mjuhasz** [comment]
The "Don't show this again" checkbox cannot be unchecked:

screen shot 2016-06-30 at 7 38 30 pm

**Comment by mattsherman** [comment]
Trying an alternate approach to showing the first view in a separate PR: https://github.com/Automattic/wp-calypso/pull/6475

**Comment by mattsherman** [comment]
Based on suggestions by @mtias and @ehg (thanks!), took an approach that simplifies things by getting rid of the higher order component approach and makes use of Redux to control the state.

I pulled in a modified version of the Redux work being done by @jonathansadowski  for https://github.com/Automattic/wp-calypso/pull/6252 (I will comment on that PR later to suggest that changes I made in this PR). 

There's still some cleanup to go and failing CircleCI tests, but this should give a feel of where things are headed.

File: client/state/first-view/reducer.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+} from 'state/action-types';
+
+const debug = debugFactory( 'calypso:first-view' );
+
+export function firstView( state = { disabled: [], visible: [] }, action ) {
+	switch ( action.type ) {
+
+		case DESERIALIZE: {
+			// only 'disabled' state is persisted
+			const newState = omit( state, 'visible', 'lastVisible' );
```
Comment (mattsherman) [inline client/state/first-view/reducer.js]: I'll remove `lastVisible` -- that is a remnant of an earlier implementation.

**Comment by rickybanister** [comment]
This is off to a nice start, excited to see it progress!

File: client/components/main/index.jsx [hunk 7]
```diff
@@ -3,17 +3,36 @@
...
 		return (
-			<main className={ classnames( this.props.className, 'main' ) } role="main">
+			<main className={ classes } role="main">
 				{ this.props.children }
 			</main>
 		);
 	}
 } );
+
+export default connect(
```
Comment (mattsherman) [inline client/components/main/index.jsx]: @ehg This approach works well. However, it requires me to modify every place that uses `<Main>` to wrap it in a `<Provider>` -- is that just the way it has to be, or do you have any suggestions for avoiding this?

For example, `/help` fails with the following:

```
Uncaught Invariant Violation: Could not find "store" in either the context or props of "Connect(Main)". Either wrap the root component in a <Provider>, or explicitly pass "store" as a prop to "Connect(Main)".
```

File: client/state/first-view/selectors.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+export function switchedToDifferentSection( state ) {
+	const section = state.ui.section;
+	const routeSets = filter( getActionLog( state ), entry => entry.type === 'ROUTE_SET' );
+	const lastRouteSetsForSection = takeRightWhile( routeSets,
+		routeSet => some( section.paths, path => startsWith( routeSet.path, path ) ) );
+	return lastRouteSetsForSection.length === 1;
+}
+
+export function shouldViewBeVisible( state ) {
```
Comment (mattsherman) [inline client/state/first-view/selectors.js]: @ehg @mcsf How does this approach look, compared to the previous page handler approach? Any further suggestions?

**Comment by mattsherman** [comment]
@jonathansadowski I tried out some changes to the Redux state in this PR, based on our conversations with @ehg and @mcsf the other day. If we agree that this is the approach to take we should fold these changes into your PR.

File: client/components/main/index.jsx (hunk 7 above)
Comment (ehg) [inline client/components/main/index.jsx]: I think there's no avoiding this until we have a single React tree. At the moment we're unmounting and mounting React trees into `#primary`, so wrapping is necessary :(

File: client/components/first-view/index.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import RootChild from 'components/root-child';
+import { getSectionName } from 'state/ui/selectors';
+import { shouldViewBeVisible } from 'state/first-view/selectors';
+import * as FirstViewActions from 'state/first-view/actions';
```
Comment (ehg) [inline client/components/first-view/index.js]: Minor, but thought I'd mention it's good to get into the habit of importing explicitly, so we can reap the reduced chunk size benefits of either the babel tree shaking plugin, or webpack2.

File: client/components/main/index.jsx (hunk 7 above)
Comment (mtias) [inline client/components/main/index.jsx]: Yes, this is something we need to go through. Perhaps do this in a separate PR so people can more easily test?

File: client/my-sites/stats/controller.js [hunk 10]
```diff
@@ -707,18 +713,20 @@ module.exports = {
...
-					siteId: siteId,
-					page: pageNum,
-					perPage: 20,
-					total: 10,
-					followersList: followersList,
-					followType: followType,
-					followList: followList,
-					domain: siteDomain
-				} ),
+				React.createElement( ReactRedux.Provider, { store: context.store },
```
Comment (ehg) [inline client/my-sites/stats/controller.js]: I forgot we have [`renderWithReduxStore`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/react-helpers/index.js#L8) for this.

File: client/components/first-view/index.js (hunk 9 above)
Comment (mattsherman) [inline client/components/first-view/index.js]: Good point, I'll be sure to do that.

File: client/components/main/index.jsx (hunk 7 above)
Comment (mattsherman) [inline client/components/main/index.jsx]: Will definitely do -- I'll create a separate PR.

File: client/my-sites/stats/controller.js (hunk 10 above)
Comment (mattsherman) [inline client/my-sites/stats/controller.js]: Ah, thanks! I wasn't aware.

File: client/state/first-view/selectors.js (hunk 8 above)
Comment (mattsherman) [inline client/state/first-view/selectors.js]: Thanks taking a look @ehg! Much appreciated!

**Comment by mattsherman** [comment]
Note: In order to iterate more quickly, I am introducing a temporary `<StatsMain>` component so that `<Main>` and all its usage doesn't need to be updated before we can ship this (#6651 will take care of this).

File: client/my-sites/stats/overview.jsx [hunk 11]
```diff
@@ -90,19 +90,22 @@ export default React.createClass( {
...
+			</StatsMain>
 		);
 	},
 
 	placeholders() {
 		const items = [];
 		const limit = Math.min( this.props.user.get().visible_site_count, 10 );
 
+		// TODO: a separate StatsSectionTitle component should be created
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Comment (mcsf) [inline client/my-sites/stats/overview.jsx]: You can use line-wise eslint disable directives, e.g.
```js
foo; // eslint-disable-line wpcalypso/jsx-classname-namespace
```

File: client/my-sites/stats/stats-main/index.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
+/** @ssr-ready **/
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { connect }	from 'react-redux';
```
Comment (mcsf) [inline client/my-sites/stats/stats-main/index.js]: Sneaky tab here :)

File: client/my-sites/stats/stats-main/index.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+import classnames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import Main from 'components/main';
+import { shouldViewBeVisible } from 'state/first-view/selectors';
+
+const StatsMain = React.createClass( {
+	displayName: 'StatsMain',
```
Comment (mcsf) [inline client/my-sites/stats/stats-main/index.js]: `displayName` shouldn't be necessary, as we're assigning the class directly to the `StatsMain` constant. Also, since this is a render-only component, why not just define it as a pure function?

```js
const StatsMain = ( { isFirstViewVisible, children } ) => {
  const classes = classnames( … );
  return (
    <Main … >
      …
    </Main>
  );
};
```

File: client/state/first-view/reducer.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+import omit from 'lodash/omit';
+import union from 'lodash/union';
```
Comment (mcsf) [inline client/state/first-view/reducer.js]: FYI, this can now be flatten into

```js
import { omit, union } from 'lodash';
```

\o/

File: client/state/first-view/reducer.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+		case FIRST_VIEW_SHOW:
+			return Object.assign( {}, state, {
+				hidden: state.hidden.filter( view => view !== action.view ),
+			} );
+
+		// This is necessary to show the first view again when we go back to a section in a single session
+		case ROUTE_SET:
+			return Object.assign( {}, state, {
+				hidden: []
+			} );
```
Comment (mcsf) [inline client/state/first-view/reducer.js]: You might also like to hear that we can drop `assign` and use pure ES6+ spread syntax:

```js
return { ...state, hidden: [] };
```

That said, there's no imperative to adopt this pattern; it's really just convenient syntax when so desired.

**Comment by mcsf** [comment]
> Note: In order to iterate more quickly, I am introducing a temporary `<StatsMain>` component so that `<Main>` and all its usage doesn't need to be updated before we can ship this (#6651 will take care of this).

Sounds sensible!

Code looks good! 👍

**Comment by mjuhasz** [comment]
Thank you for all the feedback we got here!

We are closing this PR in favor of https://github.com/Automattic/wp-calypso/pull/6717 - we have ported this code over there, and will keep refining and integrating it with the Redux components.

Please follow the further development of this code in https://github.com/Automattic/wp-calypso/pull/6717, and keep the comments coming!
